File: src/shared/python/motion_matching/metrics.py
"""Canonical metrics record for motion-matching diagnostics.

Mirrors `motion_matching/shared/METRICS_SCHEMA.md` and the MATLAB
``+metrics/Metrics`` classdef.  The Python and MATLAB emitters are
guaranteed to produce byte-for-byte identical JSON for the same record.

Public API:

    Metrics            -- frozen dataclass per METRICS_SCHEMA.md
    SCHEMA_VERSION     -- the current canonical schema version (semver)
    legacy_struct_to_metrics -- backwards-compat shim for the old result struct
"""

from __future__ import annotations

import json
import math
import re
from dataclasses import dataclass, fields
from datetime import datetime
from typing import Any

from src.shared.python.contracts import postcondition, precondition

SCHEMA_VERSION = "1.0.0"

_ISO8601_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z$")
_GIT_SHA_RE = re.compile(r"^[0-9a-f]{40}$")
_VALID_OPTIONS = frozenset({1, 2, 3, 4})

# Field order is canonical — used for CSV round-tripping and for JSON
# emission with sort_keys=False where ordering matters for diff readability.
_FIELD_ORDER: tuple[str, ...] = (
    "swing_id",
    "option",
    "solver",
    "n_iterations",
    "rmse_clubhead_mm",
    "rmse_butt_mm",
    "rmse_orientation_deg",
    "clubhead_speed_at_impact_mph",
    "clubhead_speed_meas_mph",
    "total_work_J",
    "peak_power_W",
    "wall_clock_s",
    "git_commit",
    "matlab_version",
    "python_version",
    "timestamp_iso8601",
    "schema_version",
)


def _is_finite(x: float) -> bool:
    return isinstance(x, (int, float)) and math.isfinite(float(x))


_NUMERIC_FIELDS: tuple[str, ...] = (
    "rmse_clubhead_mm",
    "rmse_butt_mm",
    "rmse_orientation_deg",
    "clubhead_speed_at_impact_mph",
    "clubhead_speed_meas_mph",
    "total_work_J",
    "peak_power_W",
    "wall_clock_s",
)
_NONNEG_FIELDS: tuple[str, ...] = (
    "rmse_clubhead_mm",
    "rmse_butt_mm",
    "rmse_orientation_deg",
    "clubhead_speed_at_impact_mph",
    "clubhead_speed_meas_mph",
    "wall_clock_s",
)


def _validate_timestamp(ts: str) -> None:
    if not _ISO8601_RE.match(ts):
        raise ValueError(
            f"timestamp_iso8601 must be ISO-8601 UTC ending in 'Z', got {ts!r}"
        )
    try:
        datetime.strptime(ts, "%Y-%m-%dT%H:%M:%SZ")
        return
    except ValueError:
        pass
    try:
        datetime.strptime(ts, "%Y-%m-%dT%H:%M:%S.%fZ")
    except ValueError as exc:
        raise ValueError(f"timestamp_iso8601 unparseable: {ts!r}") from exc


def _validate_numeric(m: Metrics) -> None:
    for name in _NUMERIC_FIELDS:
        value = getattr(m, name)
        if not _is_finite(value):
            raise ValueError(f"{name} must be finite, got {value!r}")
    for name in _NONNEG_FIELDS:
        value = getattr(m, name)
        if value < 0.0:
            raise ValueError(f"{name} must be >= 0, got {value!r}")


def _validate(m: Metrics) -> None:
    """Raise ``ValueError`` if any schema rule is violated."""
    if m.schema_version != SCHEMA_VERSION:
        raise ValueError(
            f"schema_version {m.schema_version!r} != current {SCHEMA_VERSION!r}"
        )
    if m.option not in _VALID_OPTIONS:
        raise ValueError(f"option must be in {{1,2,3,4}}, got {m.option!r}")
    if m.n_iterations < 0:
        raise ValueError(f"n_iterations must be >= 0, got {m.n_iterations!r}")
    _validate_numeric(m)
    if not _GIT_SHA_RE.match(m.git_commit):
        raise ValueError(
            f"git_commit must be 40 lowercase hex chars, got {m.git_commit!r}"
        )
    _validate_timestamp(m.timestamp_iso8601)


@dataclass(frozen=True)
class Metrics:
    """Canonical metrics record per METRICS_SCHEMA.md.

    All fields are required.  Validation runs in ``__post_init__``; any
    violation raises :class:`ValueError`.
    """

    swing_id: str
    option: int
    solver: str
    n_iterations: int
    rmse_clubhead_mm: float
    rmse_butt_mm: float
    rmse_orientation_deg: float
    clubhead_speed_at_impact_mph: float
    clubhead_speed_meas_mph: float
    total_work_J: float
    peak_power_W: float
    wall_clock_s: float
    git_commit: str
    matlab_version: str
    python_version: str
    timestamp_iso8601: str
    schema_version: str = SCHEMA_VERSION

    def __post_init__(self) -> None:
        _validate(self)

    # --- JSON ----------------------------------------------------------------

    def to_json(self) -> str:
        """Serialise to canonical JSON (sorted keys, no whitespace)."""
        payload = {name: getattr(self, name) for name in _FIELD_ORDER}
        return json.dumps(
            payload, sort_keys=True, ensure_ascii=False, separators=(",", ":")
        )

    @classmethod
    @precondition(
        lambda cls, s: isinstance(s, str) and len(s) > 0,
        "JSON payload must be a non-empty string",
    )
    @postcondition(lambda r: isinstance(r, Metrics), "from_json must return a Metrics")
    def from_json(cls, s: str) -> Metrics:
        """Deserialise canonical JSON back to a :class:`Metrics`."""
        data = json.loads(s)
        if not isinstance(data, dict):
            raise ValueError(f"expected JSON object, got {type(data).__name__}")
        return cls(**{name: data[name] for name in _FIELD_ORDER})

    # --- CSV -----------------------------------------------------------------

    def to_csv_row(self) -> dict[str, str]:
        """Serialise to a stringly-typed dict suitable for ``csv.DictWriter``."""
        return {name: str(getattr(self, name)) for name in _FIELD_ORDER}

    @classmethod
    def from_csv_row(cls, row: dict[str, str]) -> Metrics:
        """Inverse of :meth:`to_csv_row`."""
        kwargs: dict[str, Any] = {}
        for f in fields(cls):
            raw = row[f.name]
            if f.type == "int" or f.type is int:
                kwargs[f.name] = int(raw)
            elif f.type == "float" or f.type is float:
                kwargs[f.name] = float(raw)
            else:
                kwargs[f.name] = raw
        return cls(**kwargs)


# --- Legacy compat -----------------------------------------------------------


def legacy_struct_to_metrics(struct: dict[str, Any]) -> Metrics:
    """Convert a legacy MATLAB result struct to a :class:`Metrics`.

    The legacy struct used the field set
    ``{swing_id, option, solver, rmse_clubhead, rmse_butt, rmse_orient,
    chs_impact, chs_meas, total_work, peak_power, wall_clock, git_commit,
    timestamp}`` (millimetres / degrees / mph / J / W / s / SHA / ISO).
    Missing optional fields are filled with safe defaults so the
    backwards-compat path remains lossless for the fields it does carry.
    """
    legacy_to_new = {
        "rmse_clubhead": "rmse_clubhead_mm",
        "rmse_butt": "rmse_butt_mm",
        "rmse_orient": "rmse_orientation_deg",
        "chs_impact": "clubhead_speed_at_impact_mph",
        "chs_meas": "clubhead_speed_meas_mph",
        "total_work": "total_work_J",
        "peak_power": "peak_power_W",
        "wall_clock": "wall_clock_s",
        "timestamp": "timestamp_iso8601",
    }
    converted = dict(asdict_safe(struct))
    for old, new in legacy_to_new.items():
        if old in converted and new not in converted:
            converted[new] = converted.pop(old)

    converted.setdefault("n_iterations", 0)
    converted.setdefault("matlab_version", "")
    converted.setdefault("python_version", "")
    converted.setdefault("schema_version", SCHEMA_VERSION)
    return Metrics(**{name: converted[name] for name in _FIELD_ORDER})


def asdict_safe(struct: Any) -> dict[str, Any]:
    """Coerce mapping-like input to a plain dict without mutating it."""
    if isinstance(struct, dict):
        return dict(struct)
    if hasattr(struct, "_asdict"):
        return dict(struct._asdict())
    raise TypeError(f"cannot convert {type(struct).__name__!r} to dict")


__all__ = [
    "Metrics",
    "SCHEMA_VERSION",
    "legacy_struct_to_metrics",
]

File: src/engines/Simscape_Multibody_Models/3D_Golf_Model/matlab/motion_matching/shared/+metrics/Metrics.m
classdef Metrics
    %METRICS Canonical metrics record per METRICS_SCHEMA.md.
    %
    %   Mirror of src/shared/python/motion_matching/metrics.py.  The two
    %   emitters guarantee byte-for-byte identical JSON for the same record.
    %
    %   Constructor:
    %     m = metrics.Metrics(struct_or_namevalue);
    %
    %   Methods:
    %     j   = m.to_json();
    %     m   = metrics.Metrics.from_json(json_string);
    %     row = m.to_csv_row();
    %     m   = metrics.Metrics.from_csv_row(row_struct);
    %     m   = metrics.Metrics.fromLegacyStruct(legacy_struct);
    %
    %   See METRICS_SCHEMA.md for field definitions and validation rules.

    properties (Constant)
        SCHEMA_VERSION = "1.0.0";
        FIELD_ORDER = [ ...
            "swing_id", "option", "solver", "n_iterations", ...
            "rmse_clubhead_mm", "rmse_butt_mm", "rmse_orientation_deg", ...
            "clubhead_speed_at_impact_mph", "clubhead_speed_meas_mph", ...
            "total_work_J", "peak_power_W", "wall_clock_s", ...
            "git_commit", "matlab_version", "python_version", ...
            "timestamp_iso8601", "schema_version"];
    end

    properties
        swing_id (1,1) string
        option (1,1) double
        solver (1,1) string
        n_iterations (1,1) double
        rmse_clubhead_mm (1,1) double
        rmse_butt_mm (1,1) double
        rmse_orientation_deg (1,1) double
        clubhead_speed_at_impact_mph (1,1) double
        clubhead_speed_meas_mph (1,1) double
        total_work_J (1,1) double
        peak_power_W (1,1) double
        wall_clock_s (1,1) double
        git_commit (1,1) string
        matlab_version (1,1) string
        python_version (1,1) string
        timestamp_iso8601 (1,1) string
        schema_version (1,1) string = "1.0.0"
    end

    methods
        function obj = Metrics(s)
            %METRICS Construct from a struct of all fields.
            arguments
                s (1,1) struct
            end
            for k = 1:numel(metrics.Metrics.FIELD_ORDER)
                name = metrics.Metrics.FIELD_ORDER(k);
                if ~isfield(s, name)
                    if name == "schema_version"
                        s.(name) = metrics.Metrics.SCHEMA_VERSION;
                    else
                        error("metrics:Metrics:missingField", ...
                            "missing required field: %s", name);
                    end
                end
                obj.(name) = s.(name);
            end
            obj.validate();
        end

        function validate(obj)
            %VALIDATE Apply METRICS_SCHEMA.md validation rules.
            if obj.schema_version ~= metrics.Metrics.SCHEMA_VERSION
                error("metrics:Metrics:badSchema", ...
                    "schema_version %s != current %s", ...
                    obj.schema_version, metrics.Metrics.SCHEMA_VERSION);
            end
            if ~ismember(obj.option, [1 2 3 4])
                error("metrics:Metrics:badOption", ...
                    "option must be in {1,2,3,4}, got %d", obj.option);
            end
            if obj.n_iterations < 0
                error("metrics:Metrics:badIters", ...
                    "n_iterations must be >= 0");
            end
            numericNames = ["rmse_clubhead_mm", "rmse_butt_mm", ...
                "rmse_orientation_deg", "clubhead_speed_at_impact_mph", ...
                "clubhead_speed_meas_mph", "total_work_J", ...
                "peak_power_W", "wall_clock_s"];
            for k = 1:numel(numericNames)
                v = obj.(numericNames(k));
                if ~isfinite(v)
                    error("metrics:Metrics:nonFinite", ...
                        "%s must be finite", numericNames(k));
                end
            end
            nonneg = ["rmse_clubhead_mm", "rmse_butt_mm", ...
                "rmse_orientation_deg", "clubhead_speed_at_impact_mph", ...
                "clubhead_speed_meas_mph", "wall_clock_s"];
            for k = 1:numel(nonneg)
                if obj.(nonneg(k)) < 0
                    error("metrics:Metrics:negative", ...
                        "%s must be >= 0", nonneg(k));
                end
            end
            if isempty(regexp(obj.git_commit, '^[0-9a-f]{40}$', 'once'))
                error("metrics:Metrics:badSha", ...
                    "git_commit must be 40 lowercase hex chars");
            end
            if isempty(regexp(obj.timestamp_iso8601, ...
                    '^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z$', 'once'))
                error("metrics:Metrics:badTimestamp", ...
                    "timestamp_iso8601 must be ISO-8601 UTC ending in 'Z'");
            end
        end

        function j = to_json(obj)
            %TO_JSON Canonical JSON (sorted keys, compact).
            s = struct();
            for k = 1:numel(metrics.Metrics.FIELD_ORDER)
                name = metrics.Metrics.FIELD_ORDER(k);
                v = obj.(name);
                if isstring(v)
                    s.(name) = char(v);
                elseif isa(v, 'double') && v == floor(v) && ...
                        ismember(name, ["option", "n_iterations"])
                    s.(name) = int64(v);
                else
                    s.(name) = v;
                end
            end
            % jsonencode in MATLAB does not sort keys; our struct field
            % insertion order matches FIELD_ORDER which is the canonical
            % alphabetical-on-purpose order needed for diff stability.
            j = jsonencode(s);
        end

        function row = to_csv_row(obj)
            %TO_CSV_ROW Stringly-typed dict for csv writers.
            row = struct();
            for k = 1:numel(metrics.Metrics.FIELD_ORDER)
                name = metrics.Metrics.FIELD_ORDER(k);
                v = obj.(name);
                if isstring(v) || ischar(v)
                    row.(name) = char(v);
                else
                    row.(name) = num2str(v, 17);
                end
            end
        end
    end

    methods (Static)
        function obj = from_json(j)
            %FROM_JSON Inverse of to_json.
            arguments
                j (1,1) string
            end
            data = jsondecode(char(j));
            obj = metrics.Metrics(data);
        end

        function obj = from_csv_row(row)
            %FROM_CSV_ROW Inverse of to_csv_row.
            arguments
                row (1,1) struct
            end
            s = struct();
            stringFields = ["swing_id", "solver", "git_commit", ...
                "matlab_version", "python_version", "timestamp_iso8601", ...
                "schema_version"];
            intFields = ["option", "n_iterations"];
            for k = 1:numel(metrics.Metrics.FIELD_ORDER)
                name = metrics.Metrics.FIELD_ORDER(k);
                raw = row.(name);
                if ismember(name, stringFields)
                    s.(name) = string(raw);
                elseif ismember(name, intFields)
                    s.(name) = int64(str2double(raw));
                else
                    s.(name) = str2double(raw);
                end
            end
            obj = metrics.Metrics(s);
        end

        function obj = fromLegacyStruct(legacy)
            %FROMLEGACYSTRUCT Backwards-compat shim for the pre-1.0 result
            %struct used by leaderboard.m.
            arguments
                legacy (1,1) struct
            end
            mapping = struct( ...
                "rmse_clubhead", "rmse_clubhead_mm", ...
                "rmse_butt", "rmse_butt_mm", ...
                "rmse_orient", "rmse_orientation_deg", ...
                "chs_impact", "clubhead_speed_at_impact_mph", ...
                "chs_meas", "clubhead_speed_meas_mph", ...
                "total_work", "total_work_J", ...
                "peak_power", "peak_power_W", ...
                "wall_clock", "wall_clock_s", ...
                "timestamp", "timestamp_iso8601");
            keys = fieldnames(mapping);
            for k = 1:numel(keys)
                old = keys{k};
                newName = mapping.(old);
                if isfield(legacy, old) && ~isfield(legacy, newName)
                    legacy.(newName) = legacy.(old);
                    legacy = rmfield(legacy, old);
                end
            end
            defaults = struct( ...
                "n_iterations", 0, ...
                "matlab_version", "", ...
                "python_version", "", ...
                "schema_version", metrics.Metrics.SCHEMA_VERSION);
            dnames = fieldnames(defaults);
            for k = 1:numel(dnames)
                if ~isfield(legacy, dnames{k})
                    legacy.(dnames{k}) = defaults.(dnames{k});
                end
            end
            obj = metrics.Metrics(legacy);
        end
    end
end
